services/offline/src/lib/dhis2-connection-status.tsx

```diff
@@ -0,0 +1,201 @@
+import { useDataEngine } from '@dhis2/app-service-data'
+import PropTypes from 'prop-types'
+import React, {
+    useCallback,
+    useState,
+    useRef,
+    useMemo,
+    useEffect,
+    useContext,
+} from 'react'
+import { useOfflineInterface } from './offline-interface'
+import SmartInterval from './smart-interval'
+
+// Utils for saving 'last connected' datetime in local storage
+const lastConnectedKey = 'dhis2.lastConnected'
+const updateLastConnected = () => {
+    localStorage.setItem(lastConnectedKey, new Date(Date.now()).toUTCString())
+}
+const getLastConnected = () => {
+    const lastConnected = localStorage.getItem(lastConnectedKey)
+    return lastConnected ? new Date(lastConnected) : null
+}
+
+// todo: maybe make a server-health endpoint
+const pingQuery = { ping: { resource: 'system/ping' } }
+
+export interface Dhis2ConnectionStatus {
+    isConnected: boolean
+    isDisconnected: boolean
+    lastConnected: Date | null
+}
+
+const Dhis2ConnectionStatusContext = React.createContext({
+    isConnected: true,
+    isDisconnected: false,
+    lastConnected: null,
```

Comment:
I'm wondering to what extend `isDisconnected` has added value. The only advantage I could think of that this additional boolean could provide, is if you wanted to distinguish between:
1. A disconnected app that has been online at some point: `isConnected: false` and `isDisconnected: true`
2. An app that wasn't able to establish a connection from when it loaded: `isConnected: false` and `isDisconnected: false`

I initially thought that those two scenarios could also be distinguished from one another by using the `lastConnected` property, but I realise now that's not true because that field is saved to local storage and thus doesn't just apply to the current session.

I'm starting to assume quite a bit already, but if the above is correct, then perhaps the name of the `isDisconnected` field should be changed. Something like `wasConnectedDuringCurrentSession` or sth would express better what this boolean represents.

Let's first discuss if and why we need a second boolean, and then, based on the potential use-cases, discuss what would be an accurate name for it.

Reply:
For the previous online status hook, the analytics team asked for an `offline` boolean as as a convenience value intended to improve code legibility, e.g `disabled={offline}` or `ready={online}`, so I'm guessing they'll want the same for the connection status. It seems like most of their use cases are checking the `offline`/`isDisconnected` connection, for what it's worth

I think it's a simple enough thing to add since it's just a negation of the `isConnected` status, and I agree that it makes the code a little more self-descriptive, which I always like 🙂 Do you think that's good enough reason?

Reply:
Yeah, I realised later on during the review process that `isDisconnected` is simply the inverse of `isConnected`. The fact that I was confused illustrates the issue I personally had with this variable name. When thinking about the word _disconnected_ in the context of a connection to the server I made an association that perhaps I wouldn't have done if I was a native English speaker. Specifically, I thought that it suggested the user was connected at one point in time, but the connection was lost. But I've now looked up the definition of the word _disconnected_, and it literally means _not connected_, so it does not imply there was a connection at some point in time. As such I think it's fine to return both `isConnected` and `isDisconnected` as you're doing now.